### PR TITLE
EGL/X11 (Wayland, pt 1)

### PR DESCRIPTION
**Pull request by GeorgesStavracas**
This pull requests builds on top of @w23 and @cyclopsian's work, and their code (or code heavily inspired by them) is appropriately credited.

### Description

Add an alternative EGL/X11 winsys to `libobs-opengl`. Run with `OBS_USE_EGL` to use it.

### Motivation and Context

This is the 1st step of the [RFC #14](https://github.com/obsproject/rfcs/pull/14/files).

The ultimate goal of this pull request is introduce EGL in a controlled environment, in such a way that can be easily tested. In the future, another EGL winsys will be added for Wayland.

### How Has This Been Tested?

Under any desktop environment, X11 or Wayland, run:

`$ OBS_USE_EGL=1 obs`

My particular setup is Linux + GNOME + Wayland, but I also tested on X11.

### Types of changes

 - New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.

### CODE REVIEW

**Comment by GeorgesStavracas** [comment]
Changes:

 * Synchronized some parameters between GLX and EGL/X11
 * Rebased against latest master

File: libobs-opengl/gl-nix-egl-x11.c [hunk 1]
```diff
@@ -0,0 +1,689 @@
...
+	EGLContext context = EGL_NO_CONTEXT;
+	int egl_min = 0, egl_maj = 0;
+	bool success = false;
+	const char *egl_client_extensions = NULL;
+
+	eglBindAPI(EGL_OPENGL_API);
+
+	egl_client_extensions = eglQueryString(EGL_NO_DISPLAY, EGL_EXTENSIONS);
+
+	// FIXME glad
```
Comment (kkartaltepe) [inline libobs-opengl/gl-nix-egl-x11.c]: Please fix?

Comment (GeorgesStavracas) [reply libobs-opengl/gl-nix-egl-x11.c]: I'm trying to figure that one out. GLAD doesn't generate the functions for that. I don't think it's acceptable to hand edit the generated files and add it. Either we keep it like that, or edit the file. Opinions?

Comment (mufunyo) [reply libobs-opengl/gl-nix-egl-x11.c]: This should probably be a draft PR if you're still working out the kinks.

**Comment by GeorgesStavracas** [comment]
Stress tested on GNOME + Wayland, GNOME + X11, and Weston.

**Comment by Sunderland93** [comment]
Is it ready to merge?

**Comment by kkartaltepe** [comment]
> Is it ready to merge?

Not really, but its certainly ready to test. Please let us know if you run into any issues using it so we can get it merged more quickly.

**Comment by GeorgesStavracas** [comment]
> Is it ready to merge?

What @kkartaltepe said. Also, unless you want to specifically try the 3 steps individually (which is absolutely appreciated as well), please use the [third and final pull request](https://github.com/obsproject/obs-studio/pull/2484). Thanks for testing it!

**Comment by kode54** [comment]
Will this be gaining a method to fix the Preview canvas? That seems to be broken here, even with this patch.

**Comment by Fenrirthviti** [comment]
There are two additional Wayland PRs, I would recommend looking at them all.

**Comment by kode54** [comment]
@Fenrirthviti I have no idea how I ended up replying to part 1 instead of 3. I applied part 3 as a patch, after rebasing it, and fighting with it because I didn’t realize that in its current state, generating a patch from pull part 2 includes part 1, and part 3 includes parts 1 and 2 already.

**Pull request merged**